File: extensions/redditisfun/src/main/java/app/morphe/extension/redditisfun/FixRedgifsApiPatch.java
/*
 * Copyright 2026 wchill.
 * https://github.com/wchill/patcheddit
 *
 * See the included NOTICE file for GPLv3 §7(b) and §7(c) terms that apply to this code.
 */

package app.morphe.extension.redditisfun;

import app.morphe.extension.shared.Logger;
import app.morphe.extension.shared.fixes.redgifs.RedgifsTokenManager;

/**
 * @noinspection unused
 */
public class FixRedgifsApiPatch {
    // RIF's hardcoded RedGifs user agent; RedGifs binds the temporary token to it.
    private static final String USER_AGENT = "org.quantumbadger.redreader/1.25.1";

    public static String getRedgifsToken() {
        try {
            return RedgifsTokenManager.refreshToken(USER_AGENT).getAccessToken();
        } catch (Exception ex) {
            Logger.printException(() -> "Could not fetch RedGifs token", ex);
            return null;
        }
    }
}

File: patches/src/main/kotlin/app/morphe/patches/reddit/customclients/redditisfun/fix/redgifs/FixRedgifsApiPatch.kt
/*
 * Copyright 2026 wchill.
 * https://github.com/wchill/patcheddit
 *
 * See the included NOTICE file for GPLv3 §7(b) and §7(c) terms that apply to this code.
 */

package app.morphe.patches.reddit.customclients.redditisfun.fix.redgifs

import app.morphe.patcher.extensions.InstructionExtensions.addInstructions
import app.morphe.patcher.patch.bytecodePatch
import app.morphe.patches.reddit.customclients.AppCompatibility
import app.morphe.patches.reddit.customclients.ExtensionPatches

private const val EXTENSION_CLASS_DESCRIPTOR =
    "Lapp/morphe/extension/redditisfun/FixRedgifsApiPatch;"

val fixRedgifsApiPatch = bytecodePatch(
    name = "Fix Redgifs API",
    description = "Fix RedGifs videos and gifs not loading.",
    default = true,
) {
    dependsOn(ExtensionPatches.RIF)
    compatibleWith(*AppCompatibility.RedditIsFun)

    execute {
        // RIF requests a token from the removed /v2/oauth/client endpoint. Return a valid
        // RedGifs temporary token from the extension instead.
        RedgifsTokenFingerprint.method.addInstructions(
            0,
            """
                invoke-static { }, $EXTENSION_CLASS_DESCRIPTOR->getRedgifsToken()Ljava/lang/String;
                move-result-object v0
                return-object v0
            """,
        )
    }
}
